Initial build on x86-64 fails. Works after manually invoking 'make codegen'

Initial configuration and building process:

~~~
foo@bar ~/projekte/openage (git)-[master] % ./configure --mode=debug --compiler=clang
./configure is a convenience script:
it creates the build directory,  symlinks it,
and invokes cmake for an out-of-source build.

Nobody is stopping you from skipping ./configure and our Makefile,
and using CMake directly (e.g. when packaging, or using an IDE).
For your convenience, ./configure even prints the direct CMake invocation!

         build_type | Debug
       cxx_compiler | clang++
          cxx_flags | 
   exe_linker_flags | 
     install_prefix | /usr/local
module_linker_flags | 
shared_linker_flags | 

config options:

          backtrace | if_available
gperftools-profiler | if_available
gperftools-tcmalloc | False
            inotify | if_available

bindir:
/home/foo/projekte/openage/.bin/clang++-debug-Oauto-sanitize-none/

invocation:
cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_CXX_FLAGS='' -DCMAKE_EXE_LINKER_FLAGS='' -DCMAKE_INSTALL_PREFIX=/usr/local -DCMAKE_MODULE_LINKER_FLAGS='' -DCMAKE_SHARED_LINKER_FLAGS='' -DCXX_OPTIMIZATION_LEVEL=auto -DCXX_SANITIZE_FATAL=False -DCXX_SANITIZE_MODE=none -DWANT_BACKTRACE=if_available -DWANT_GPERFTOOLS_PROFILER=if_available -DWANT_GPERFTOOLS_TCMALLOC=False -DWANT_INOTIFY=if_available -- /home/foo/projekte/openage

(running cmake...)

 ___  ______ _______ _______ ___
|  _)/ _____|_______|_______|_  |
| | ( (____  _____      _     | |    ___  ____  _____ ____  _____  ____ _____
| |  \____ \|  ___)    | |    | |   / _ \|  _ \| ___ |  _ \(____ |/ _  | ___ |
| |_ _____) ) |        | |   _| |  | |_| | |_| | ____| | | / ___ ( (_| | ____|
|___|______/|_|        |_|  (___|   \___/|  __/|_____)_| |_\_____|\___ |_____)
                                         |_|                     (_____|

Welcome to the SFT technologies computer-aided openage build system!

You have chosen, or been chosen, to attempt the daring task of building openage.
If you have installed all the dependencies that are conveniently listed in
[doc/building.md], this _might_ just work!

If it doesn't, consider reporting the issue/asking for help in #sfttech on freenode.net.

-- The CXX compiler identification is Clang 4.0.1
-- Check for working CXX compiler: /usr/bin/clang++
-- Check for working CXX compiler: /usr/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test HAVE_THREAD_LOCAL_STORAGE
-- Performing Test HAVE_THREAD_LOCAL_STORAGE - Failed
readline() on closed filehandle FILE at /usr/bin/python_count line 80.
readline() on closed filehandle FILE at /usr/bin/python_count line 80.
-- Dropping invalid python interpreter '/usr/bin/python_count'
-- Using python interpreter: /usr/bin/python3
-- Found Python: /usr/bin/python3 (Required is at least version "3.4") 
-- Found Cython: /usr/bin/python3 -m cython (found suitable version "0.26.1", minimum required is "0.25") 
-- Found numpy: /usr/lib/python3.6/site-packages/numpy/core/include  
-- Checking python3 module PIL.Image - Success
-- Checking python3 module PIL.ImageDraw - Success
-- Checking python3 module numpy - Success
-- Checking python3 module pygments - Success
-- Checking utility program opusenc - Success
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.13") found components:  doxygen dot 

cpp library
	libopenage                                          [240 sources] [14 generated]
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
-- Found Freetype: /usr/lib/libfreetype.so (found version "2.8.0") 
-- Found OpenGL: /usr/lib/libGL.so  
-- Found ZLIB: /usr/lib/libz.so (found version "1.2.11") 
-- Found PNG: /usr/lib/libpng.so (found version "1.6.32") 
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found SDL2: /usr/include/SDL2  
-- Found SDL2Image: /usr/lib/libSDL2_image.so  
-- Found Opus: /usr/lib/libopusfile.so  
-- Found Epoxy: /usr/lib/libepoxy.so  
-- Found HarfBuzz: /usr/lib/libharfbuzz.so (found suitable version "1.5.0", minimum required is "1.0.0") 
-- Could NOT find GCCBacktrace (missing: GCCBacktrace_LIBRARIES GCCBacktrace_INCLUDE_DIRS) 
unavailable: backtrace
-- Found Gperftools: /lib/libtcmalloc_and_profiler.so  
-- Found INOTIFY: /usr/include  

cython module
	run                                                 [embedded interpreter] [noinstall]
	openage.cabextract.lzxd
	openage.cabextract.cabchecksum                      [standalone]
	openage.convert.slp
	openage.convert.interface.visgrep
	openage.cppinterface.exctranslate
	openage.cppinterface.exctranslate_tests
	openage.cppinterface.pyobject
	openage.cppinterface.setup_checker
	openage.cvar.cvar
	openage.game.main_cpp
	openage.log.log_cpp
	openage.util.filelike.cpp
	openage.util.fslike.cpp
	openage.testing.cpp_testing
	openage.testing.misc_cpp

enabled options:
	gperftools-profiler
	inotify

disabled options:
	backtrace
	gperftools-tcmalloc

openage v0.3.0-1017-g914aea13

         compiler | Clang 4.0.1
       build type | Debug
         cxxflags | 
 build type flags | -g -O0
        build dir | /home/foo/projekte/openage/.bin/clang++-debug-Oauto-sanitize-none
   install prefix | /usr/local
py install prefix | /usr/lib/python3.6/site-packages

-- Configuring done
-- Generating done
-- Build files have been written to: /home/foo/projekte/openage/.bin/clang++-debug-Oauto-sanitize-none
foo@bar ~/projekte/openage (git)-[master] % make
Scanning dependencies of target pxdgen
[  1%] pxdgen: generating .pxd files from headers
generating /home/foo/projekte/openage/libopenage/__init__.py
generating libopenage/main.pxd
generating libopenage/engine.pxd
generating /home/foo/projekte/openage/libopenage/cvar/__init__.py
generating libopenage/cvar/cvar.pxd
generating /home/foo/projekte/openage/libopenage/error/__init__.py
generating libopenage/error/backtrace.pxd
generating libopenage/error/error.pxd
generating /home/foo/projekte/openage/libopenage/input/__init__.py
generating libopenage/input/input_manager.pxd
generating /home/foo/projekte/openage/libopenage/log/__init__.py
generating libopenage/log/level.pxd
generating libopenage/log/log.pxd
generating libopenage/log/logsource.pxd
generating libopenage/log/named_logsource.pxd
generating libopenage/log/message.pxd
generating /home/foo/projekte/openage/libopenage/pyinterface/__init__.py
generating libopenage/pyinterface/exctranslate.pxd
generating libopenage/pyinterface/exctranslate_tests.pxd
generating libopenage/pyinterface/defs.pxd
generating libopenage/pyinterface/functional.pxd
generating libopenage/pyinterface/hacks.pxd
generating libopenage/pyinterface/pyexception.pxd
generating libopenage/pyinterface/pyobject.pxd
generating libopenage/pyinterface/setup.pxd
generating /home/foo/projekte/openage/libopenage/testing/__init__.py
generating libopenage/testing/testlist.pxd
generating /home/foo/projekte/openage/libopenage/util/__init__.py
generating libopenage/util/compiler.pxd
generating libopenage/util/enum.pxd
generating libopenage/util/enum_test.pxd
generating libopenage/util/file.pxd
generating libopenage/util/path.pxd
generating /home/foo/projekte/openage/libopenage/util/compress/__init__.py
generating libopenage/util/compress/lzxd.pxd
generating /home/foo/projekte/openage/libopenage/util/filelike/__init__.py
generating libopenage/util/filelike/filelike.pxd
generating libopenage/util/filelike/python.pxd
generating /home/foo/projekte/openage/libopenage/util/fslike/__init__.py
generating libopenage/util/fslike/fslike.pxd
generating libopenage/util/fslike/python.pxd
Scanning dependencies of target libopenage_autogen
make[3]: *** No rule to make target '../../libopenage/gamedata/tech.gen.cpp', needed by 'libopenage/CMakeFiles/libopenage_autogen'.  Stop.
make[2]: *** [CMakeFiles/Makefile2:1550: libopenage/CMakeFiles/libopenage_autogen.dir/all] Error 2
make[1]: *** [Makefile:141: all] Error 2
make: *** [Makefile:33: build] Error 2
~~~
Manual invocation of 'make codegen':

~~~
foo@bar ~/projekte/openage (git)-[master] % make codegen                                                                                                                                :(
make --no-print-directory -C bin codegen
Scanning dependencies of target codegen
openage.codegen: generating c++ code
libopenage/gamedata/util.gen.h
libopenage/gamedata/sound.gen.h
libopenage/gamedata/graphic.gen.h
libopenage/gamedata/terrain.gen.h
libopenage/gamedata/tech.gen.h
libopenage/gamedata/unit.gen.h
libopenage/gamedata/civilisation.gen.h
libopenage/gamedata/research.gen.h
libopenage/gamedata/gamedata.gen.h
libopenage/gamedata/blending_mode.gen.h
libopenage/gamedata/color.gen.h
libopenage/gamedata/texture.gen.h
libopenage/gamedata/string_resource.gen.h
libopenage/gamedata/util.gen.cpp
libopenage/gamedata/sound.gen.cpp
libopenage/gamedata/graphic.gen.cpp
libopenage/gamedata/terrain.gen.cpp
libopenage/gamedata/tech.gen.cpp
libopenage/gamedata/unit.gen.cpp
libopenage/gamedata/civilisation.gen.cpp
libopenage/gamedata/research.gen.cpp
libopenage/gamedata/gamedata.gen.cpp
libopenage/gamedata/blending_mode.gen.cpp
libopenage/gamedata/color.gen.cpp
libopenage/gamedata/texture.gen.cpp
libopenage/gamedata/string_resource.gen.cpp
libopenage/testing/testlist.gen.cpp
~~~
 Now building starts fine:

~~~
foo@bar ~/projekte/openage (git)-[master] % make
[  1%] Automatic MOC for target libopenage
Generating MOC predefs moc_predefs.h
Generating MOC source EWIEGA46WW/moc_engine.cpp
. . .
~~~

--- a/assets/.gitignore
+++ b/assets/.gitignore
@@ -1,2 +1,2 @@
 # converted game assets
-/converted
+/converted/

--- a/buildsystem/codegen.cmake
+++ b/buildsystem/codegen.cmake
@@ -1,4 +1,4 @@
-# Copyright 2014-2015 the openage authors. See copying.md for legal info.
+# Copyright 2014-2017 the openage authors. See copying.md for legal info.
 
 function(codegen_run)
 	# this function must be called once all required assets have been created, but before the executable is finalized.
@@ -6,7 +6,7 @@ function(codegen_run)
 	# make sure this function gets called only once
 	get_property(codegen_run GLOBAL PROPERTY SFT_CODEGEN_HAS_BEEN_RUN)
 	if(codegen_run)
-		message(FATAL_ERROR codegen has already been run)
+		message(FATAL_ERROR "codegen has already been run")
 	endif()
 	set_property(GLOBAL PROPERTY SFT_CODEGEN_HAS_BEEN_RUN 1)
 

--- a/buildsystem/codegen.cmake
+++ b/buildsystem/codegen.cmake
@@ -29,10 +29,16 @@ function(codegen_run)
 	FILE(READ "${CMAKE_BINARY_DIR}/codegen_depends" CODEGEN_DEPENDS)
 	STRING(REGEX REPLACE "\n" ";" CODEGEN_TARGET_FILES ${CODEGEN_TARGET_FILES})
 	STRING(REGEX REPLACE "\n" ";" CODEGEN_DEPENDS ${CODEGEN_DEPENDS})
+
+	# as the codegen creates all files at once,
+	# let the buildsystem only depend on this single dummy file.
+	# otherwise the codegen invocation would be done for each generated source.
 	set(CODEGEN_TIMEFILE "${CMAKE_BINARY_DIR}/codegen_timefile")
 
+	list(APPEND CODEGEN_TARGET_FILES ${CODEGEN_TIMEFILE})
+
 	add_custom_command(
-		OUTPUT "${CODEGEN_TIMEFILE}"
+		OUTPUT ${CODEGEN_TARGET_FILES}
 		COMMAND ${CODEGEN_INVOCATION} --mode=codegen "--touch-file-on-cache-change=${CMAKE_CURRENT_LIST_FILE}" --force-rerun-on-generated-list-change
 		COMMAND "${CMAKE_COMMAND}" -E touch "${CODEGEN_TIMEFILE}"
 		WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"

--- a/libopenage/CMakeLists.txt
+++ b/libopenage/CMakeLists.txt
@@ -5,9 +5,8 @@
 # add new sources here, dependencies for linking and including
 # are specified below the source file list.
 
-set(CMAKE_AUTOMOC ON)
-
 declare_binary(libopenage)
+
 
 add_sources(libopenage
 	assetmanager.cpp
@@ -57,7 +56,11 @@ add_sources(libopenage GENERATED ${CODEGEN_TARGET_TUS})
 
 # after this point, no further sources can be added to the executable
 finalize_binary(libopenage openage library allow_no_undefined)
-set_target_properties(libopenage PROPERTIES VERSION 0)
+set_target_properties(libopenage PROPERTIES
+	VERSION 0
+	AUTOMOC ON
+	AUTOGEN_TARGET_DEPENDS "codegen"
+)
 
 # library dependency specification
 

